Configure TypeScript compiler (#3)

birthdaysgift opened the pull request:
Parent story: https://github.com/fenya123/bingin/issues/9

In the scope of adding Angular (#1) we've also added minimal typescript configuration required for angular to run the app.

As a part of adding linters configuration story (https://github.com/fenya123/bingin/issues/9) we should configure our typescript compiler checks to strict as possible to find all possible typing errors.

In the scope of this task we need to go through all available TS compiler checks from [docs](https://www.typescriptlang.org/tsconfig) and add useful ones in our `tsconfig.json`.

birthdaysgift commented on tsconfig.json:
```diff
@@ -1,14 +1,35 @@
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
     "outDir": "./dist/out-tsc",
+    "removeComments": true,
```
Эта опция отвечает за удаление комментариев в скомпилированном коде. Не уверен, имеет ли какой-то смысл оставлять эти комментарии в бандле даже в рамках dev сборки, поэтому решил их убрать.

fowesn commented on tsconfig.json:
```diff
@@ -1,14 +1,35 @@
+    "strictBindCallApply": true,
+    "strictFunctionTypes": true,
+    "strictNullChecks": true,
+    "strictPropertyInitialization": true,
```
Это хорошее ограничение для TypeScript, которое неприменимо к разработке на Angular, поскольку в Angular зачастую свойства класса инициализируются не в конструкторе, а в хуках жизненного цикла компонента, таких как `OnInit` и `AfterViewInit`. 
Примером может быть любое свойство компонента, в котором содержится ссылка на элемент вёрстки из HTML, связанного с компонентом. Синтаксис такой:
``` ts
// ts
@ViewChild('htmlElementName') htmlElementVarName = ElementRef<HTMLElement>
```
В данном примере мы говорим, что хотим в свойстве `htmlElementVarName` видеть HTML элемент из этого компонента, у которого установлен идентификатор `#htmlElementName`, т.е.:
``` html
<!-- HTML -->
<some-tag #htmlElementName>…</some-tag>
```
   
Это несколько похоже на следующее:
``` html
<!-- HTML -->
<some-tag id="htmlElementName">…</some-tag>
```
``` ts
// ts
const htmlElementVarName: HTMLElement = document.getElementById('htmlElementName');
```
Но на самом деле есть некоторые отличия. Например, мы не знаем, в какой момент HTML-элемент доступен в DOM, и, соответственно, не знаем, когда в ts-коде его инициализировать, а Angular знает (см. хук [AfterViewInit](https://angular.io/api/core/AfterViewInit)). Также мы получим здесь элемент с типом `HTMLElement`, в то время как Angular даст нам элемент с типом `ElementRef<HTMLElement>`. И если `ElementRef<T>` не особо интересен, то типизация элементов вёрстки, являющихся компонентами, уже имеет намного больше смысла, и этого мы тоже не можем сделать без Angular или в конструкторе. Например, у нас есть компонент `AComponent`, который мы хотим отрисовать в компоненте `BComponent`, это будет выглядеть так:
``` html
<!-- b.component.html-->
<app-a #aComponent>…</app-a>
```
``` ts
// b.component.ts
@ViewChild('aComponent') aComponent: AComponent;
```
То есть, элемент вёрстки, полученный таким образом, будет иметь тип `AComponent`, что позволит нам работать с предоставляемым им функционалом на 100% типизировано, в отличие от кейса, когда мы получаем этот компонент через `document.getElementById(…)`, потому что эта функция вернёт `HTMLElement`, который совершенно не знает о полях и методах класса `AComponent`, который им представлен.   

Это не единственный кейс, который вызовет проблемы при включённой опции `strictPropertyInitialization`, но один из самых наглядных. Что касается недекорированных свойств компонента, они зачастую инициализируются в хуке инициализации (см. хук [OnInit](https://angular.io/api/core/OnInit)), например, формы.

birthdaysgift replied:
Интересная статья в которой объясняются разные подходы к решению этой проблемы:
https://monsterlessons-academy.com/posts/angular-property-has-no-initializer-and-is-not-definitely-assigned-in-the-constructor

В этой статье рекомендуемой опцией для нетривиальных кейсов считается инициализация через `undefined`:
```typescript
SomeComponent | undefined = undefined;
```
но кажется это создаст большие трудности из ничего - перед каждым использованием компонента нам придётся ставить тайп гард на то, что компонент не `undefined`, но в таком случае не понятно, что мы должны сделать если он всё-таки `undefined`, т.е. не понятно, что писать в тайп гард

fowesn replied:
я думаю, тебе так делать не даст опция выше `exactOptionalPropertyTypes`, хотя не уверена

birthdaysgift commented on tsconfig.json:
```diff
@@ -1,14 +1,35 @@
 {
   "compilerOptions": {
```
Кроме опций компилятора самого тайпскрипта, в этот конфиг также можно вписать опции компилятора ангуляра в секцию `"angularCompilerOptions": {}`. Кажется, [дока](https://angular.io/guide/angular-compiler-options) говорит о том, что эти опции будут задейстованы, только если мы используем Ahead Of Time компиляцию, но на данный момент мне не совсем понятно, использую ли я AOT компиляцю во время разработки, или она важна только для прод билда приложения.

Парочка опций из этой конфигурации кажутся полезными даже во время разработки, а именно: [fullTemplateTypeCheck](https://angular.io/guide/angular-compiler-options#fulltemplatetypecheck) и [strictTemplates](https://angular.io/guide/angular-compiler-options#stricttemplates). Но опять же не понятно, будут ли эти проверки запускаться во время режима разработки, напр. `ng serve`, или только во время прод сборки.

birthdaysgift committed:
Configure TypeScript compiler (#3)

Parent story: https://github.com/fenya123/bingin/issues/9

In the scope of adding Angular (#1) we've also added minimal typescript configuration required for angular to run the app.

As a part of adding linters configuration story (https://github.com/fenya123/bingin/issues/9) we should configure our typescript compiler checks to strict as possible to find all possible typing errors.

In the scope of this task we need to go through all available TS compiler checks from [docs](https://www.typescriptlang.org/tsconfig) and add useful ones in our `tsconfig.json`.
